src/main/java/org/opensearch/security/action/apitokens/ApiTokenIndexHandler.java

```diff
@@ -55,66 +52,69 @@ public ApiTokenIndexHandler(Client client, ClusterService clusterService) {
         this.clusterService = clusterService;
     }
 
-    public void indexTokenMetadata(ApiToken token) {
+    public void indexTokenMetadata(ApiToken token, ActionListener<Void> listener) {
         try {
-
             XContentBuilder builder = XContentFactory.jsonBuilder();
             String jsonString = token.toXContent(builder, ToXContent.EMPTY_PARAMS).toString();
 
             IndexRequest request = new IndexRequest(ConfigConstants.OPENSEARCH_API_TOKENS_INDEX).source(jsonString, XContentType.JSON);
 
-            ActionListener<IndexResponse> irListener = ActionListener.wrap(idxResponse -> {
+            client.index(request, ActionListener.wrap(indexResponse -> {
                 LOGGER.info("Created {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
-            }, (failResponse) -> {
-                LOGGER.error(failResponse.getMessage());
+                listener.onResponse(null);
+            }, exception -> {
+                LOGGER.error(exception.getMessage());
                 LOGGER.info("Failed to create {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
-            });
-            client.index(request, irListener);
+                listener.onFailure(exception);
+            }));
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
-
     }
 
-    public void deleteToken(String name) throws ApiTokenException {
+    public void deleteToken(String name, ActionListener<Void> listener) {
         DeleteByQueryRequest request = new DeleteByQueryRequest(ConfigConstants.OPENSEARCH_API_TOKENS_INDEX).setQuery(
             QueryBuilders.matchQuery(NAME_FIELD, name)
         ).setRefresh(true);
 
-        BulkByScrollResponse response = client.execute(DeleteByQueryAction.INSTANCE, request).actionGet();
-
-        long deletedDocs = response.getDeleted();
-
-        if (deletedDocs == 0) {
-            throw new ApiTokenException("No token found with name " + name);
-        }
+        client.execute(DeleteByQueryAction.INSTANCE, request, ActionListener.wrap(response -> {
+            long deletedDocs = response.getDeleted();
+            if (deletedDocs == 0) {
+                listener.onFailure(new ApiTokenException("No token found with name " + name));
+            } else {
+                listener.onResponse(null);
```

Comment:
No synchronization after delete? Will this ensure that the token is no longer able to be used regardless of which node a request is made against?

Reply:
Sync happened on the action level which wraps this successful actionlistener